Build failure with `unix-2.8.*`

As spotted when building `text-show-instances` with GHC 9.6.1-alpha1:

```
[14 of 29] Compiling TextShow.System.Posix ( src/TextShow/System/Posix.hs, /home/ryanglscott/Documents/Hacking/Haskell/ci-maintenance/checkout/RyanGlScott/text-show-instances/dist-newstyle/build/x86_64-linux/ghc-9.6.0.20230111/text-show-instances-3.9.2/build/TextShow/System/Posix.o, /home/ryanglscott/Documents/Hacking/Haskell/ci-maintenance/checkout/RyanGlScott/text-show-instances/dist-newstyle/build/x86_64-linux/ghc-9.6.0.20230111/text-show-instances-3.9.2/build/TextShow/System/Posix.dyn_o )

src/TextShow/System/Posix.hs:26:27: error:
    Module ‘System.Posix.User’ does not export ‘GroupEntry’
   |
26 | import System.Posix.User (GroupEntry, UserEntry)
   |                           ^^^^^^^^^^

src/TextShow/System/Posix.hs:26:39: error:
    Module ‘System.Posix.User’ does not export ‘UserEntry’
   |
26 | import System.Posix.User (GroupEntry, UserEntry)
   |                                       ^^^^^^^^^

```

--- a/.github/workflows/haskell-ci.yml
+++ b/.github/workflows/haskell-ci.yml
@@ -8,9 +8,9 @@
 #
 # For more information, see https://github.com/haskell-CI/haskell-ci
 #
-# version: 0.14.3
+# version: 0.15.20230304
 #
-# REGENDATA ("0.14.3",["github","--config=cabal.haskell-ci","cabal.project"])
+# REGENDATA ("0.15.20230304",["github","--config=cabal.haskell-ci","cabal.project"])
 #
 name: Haskell-CI
 on:
@@ -19,7 +19,7 @@ on:
 jobs:
   linux:
     name: Haskell-CI - Linux - ${{ matrix.compiler }}
-    runs-on: ubuntu-18.04
+    runs-on: ubuntu-20.04
     timeout-minutes:
       60
     container:
@@ -28,9 +28,19 @@ jobs:
     strategy:
       matrix:
         include:
-          - compiler: ghc-9.2.2
+          - compiler: ghc-9.6.0.20230302
             compilerKind: ghc
-            compilerVersion: 9.2.2
+            compilerVersion: 9.6.0.20230302
+            setup-method: ghcup
+            allow-failure: true
+          - compiler: ghc-9.4.4
+            compilerKind: ghc
+            compilerVersion: 9.4.4
+            setup-method: ghcup
+            allow-failure: false
+          - compiler: ghc-9.2.7
+            compilerKind: ghc
+            compilerVersion: 9.2.7
             setup-method: ghcup
             allow-failure: false
           - compiler: ghc-9.0.2
@@ -76,18 +86,20 @@ jobs:
           apt-get install -y --no-install-recommends gnupg ca-certificates dirmngr curl git software-properties-common libtinfo5
           if [ "${{ matrix.setup-method }}" = ghcup ]; then
             mkdir -p "$HOME/.ghcup/bin"
-            curl -sL https://downloads.haskell.org/ghcup/0.1.17.5/x86_64-linux-ghcup-0.1.17.5 > "$HOME/.ghcup/bin/ghcup"
+            curl -sL https://downloads.haskell.org/ghcup/0.1.18.0/x86_64-linux-ghcup-0.1.18.0 > "$HOME/.ghcup/bin/ghcup"
             chmod a+x "$HOME/.ghcup/bin/ghcup"
-            "$HOME/.ghcup/bin/ghcup" install ghc "$HCVER"
-            "$HOME/.ghcup/bin/ghcup" install cabal 3.6.2.0
+            "$HOME/.ghcup/bin/ghcup" config add-release-channel https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-prereleases-0.0.7.yaml;
+            "$HOME/.ghcup/bin/ghcup" install ghc "$HCVER" || (cat "$HOME"/.ghcup/logs/*.* && false)
+            "$HOME/.ghcup/bin/ghcup" install cabal 3.9.0.0 || (cat "$HOME"/.ghcup/logs/*.* && false)
           else
             apt-add-repository -y 'ppa:hvr/ghc'
             apt-get update
             apt-get install -y "$HCNAME"
             mkdir -p "$HOME/.ghcup/bin"
-            curl -sL https://downloads.haskell.org/ghcup/0.1.17.5/x86_64-linux-ghcup-0.1.17.5 > "$HOME/.ghcup/bin/ghcup"
+            curl -sL https://downloads.haskell.org/ghcup/0.1.18.0/x86_64-linux-ghcup-0.1.18.0 > "$HOME/.ghcup/bin/ghcup"
             chmod a+x "$HOME/.ghcup/bin/ghcup"
-            "$HOME/.ghcup/bin/ghcup" install cabal 3.6.2.0
+            "$HOME/.ghcup/bin/ghcup" config add-release-channel https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-prereleases-0.0.7.yaml;
+            "$HOME/.ghcup/bin/ghcup" install cabal 3.9.0.0 || (cat "$HOME"/.ghcup/logs/*.* && false)
           fi
         env:
           HCKIND: ${{ matrix.compilerKind }}
@@ -105,20 +117,20 @@ jobs:
             echo "HC=$HC" >> "$GITHUB_ENV"
             echo "HCPKG=$HOME/.ghcup/bin/$HCKIND-pkg-$HCVER" >> "$GITHUB_ENV"
             echo "HADDOCK=$HOME/.ghcup/bin/haddock-$HCVER" >> "$GITHUB_ENV"
-            echo "CABAL=$HOME/.ghcup/bin/cabal-3.6.2.0 -vnormal+nowrap" >> "$GITHUB_ENV"
+            echo "CABAL=$HOME/.ghcup/bin/cabal-3.9.0.0 -vnormal+nowrap" >> "$GITHUB_ENV"
           else
             HC=$HCDIR/bin/$HCKIND
             echo "HC=$HC" >> "$GITHUB_ENV"
             echo "HCPKG=$HCDIR/bin/$HCKIND-pkg" >> "$GITHUB_ENV"
             echo "HADDOCK=$HCDIR/bin/haddock" >> "$GITHUB_ENV"
-            echo "CABAL=$HOME/.ghcup/bin/cabal-3.6.2.0 -vnormal+nowrap" >> "$GITHUB_ENV"
+            echo "CABAL=$HOME/.ghcup/bin/cabal-3.9.0.0 -vnormal+nowrap" >> "$GITHUB_ENV"
           fi
 
           HCNUMVER=$(${HC} --numeric-version|perl -ne '/^(\d+)\.(\d+)\.(\d+)(\.(\d+))?$/; print(10000 * $1 + 100 * $2 + ($3 == 0 ? $5 != 1 : $3))')
           echo "HCNUMVER=$HCNUMVER" >> "$GITHUB_ENV"
           echo "ARG_TESTS=--enable-tests" >> "$GITHUB_ENV"
           echo "ARG_BENCH=--enable-benchmarks" >> "$GITHUB_ENV"
-          echo "HEADHACKAGE=false" >> "$GITHUB_ENV"
+          if [ $((HCNUMVER >= 90600)) -ne 0 ] ; then echo "HEADHACKAGE=true" >> "$GITHUB_ENV" ; else echo "HEADHACKAGE=false" >> "$GITHUB_ENV" ; fi
           echo "ARG_COMPILER=--$HCKIND --with-compiler=$HC" >> "$GITHUB_ENV"
           echo "GHCJSARITH=0" >> "$GITHUB_ENV"
         env:
@@ -147,6 +159,18 @@ jobs:
           repository hackage.haskell.org
             url: http://hackage.haskell.org/
           EOF
+          if $HEADHACKAGE; then
+          cat >> $CABAL_CONFIG <<EOF
+          repository head.hackage.ghc.haskell.org
+             url: https://ghc.gitlab.haskell.org/head.hackage/
+             secure: True
+             root-keys: 7541f32a4ccca4f97aea3b22f5e593ba2c0267546016b992dfadcd2fe944e55d
+                        26021a13b401500c8eb2761ca95c61f2d625bfef951b939a8124ed12ecf07329
+                        f76d08be13e9a61a377a85e2fb63f4c5435d40f8feb3e12eb05905edb8cdea89
+             key-threshold: 3
+          active-repositories: hackage.haskell.org, head.hackage.ghc.haskell.org:override
+          EOF
+          fi
           cat >> $CABAL_CONFIG <<EOF
           program-default-options
             ghc-options: $GHCJOBS +RTS -M3G -RTS
@@ -170,7 +194,7 @@ jobs:
           chmod a+x $HOME/.cabal/bin/cabal-plan
           cabal-plan --version
       - name: checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
         with:
           path: source
       - name: initial cabal.project for sdist
@@ -200,6 +224,9 @@ jobs:
           package text-show-instances
             ghc-options: -Werror
           EOF
+          if $HEADHACKAGE; then
+          echo "allow-newer: $($HCPKG list --simple-output | sed -E 's/([a-zA-Z-]+)-[0-9.]+/*:\1,/g')" >> cabal.project
+          fi
           $HCPKG list --simple-output --names-only | perl -ne 'for (split /\s+/) { print "constraints: $_ installed\n" unless /^(text-show-instances)$/; }' >> cabal.project.local
           cat cabal.project
           cat cabal.project.local
@@ -207,8 +234,8 @@ jobs:
         run: |
           $CABAL v2-build $ARG_COMPILER $ARG_TESTS $ARG_BENCH --dry-run all
           cabal-plan
-      - name: cache
-        uses: actions/cache@v2
+      - name: restore cache
+        uses: actions/cache/restore@v3
         with:
           key: ${{ runner.os }}-${{ matrix.compiler }}-${{ github.sha }}
           path: ~/.cabal/store
@@ -229,4 +256,10 @@ jobs:
           ${CABAL} -vnormal check
       - name: haddock
         run: |
-          $CABAL v2-haddock $ARG_COMPILER --with-haddock $HADDOCK $ARG_TESTS $ARG_BENCH all
+          $CABAL v2-haddock --disable-documentation --haddock-all $ARG_COMPILER --with-haddock $HADDOCK $ARG_TESTS $ARG_BENCH all
+      - name: save cache
+        uses: actions/cache/save@v3
+        if: always()
+        with:
+          key: ${{ runner.os }}-${{ matrix.compiler }}-${{ github.sha }}
+          path: ~/.cabal/store

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,6 @@
+### next [????.??.??]
+* Allow building with `unix-2.8.*`.
+
 ### 3.9.4 [2022.12.30]
 * Allow building with `aeson-2.1.*`.
 

--- a/src/TextShow/System/Posix.hs
+++ b/src/TextShow/System/Posix.hs
@@ -23,7 +23,12 @@ module TextShow.System.Posix () where
 #if !defined(mingw32_HOST_OS)
 import System.Posix.DynamicLinker (RTLDFlags, DL)
 import System.Posix.Process (ProcessStatus)
-import System.Posix.User (GroupEntry, UserEntry)
+# if MIN_VERSION_unix(2,8,0)
+import System.Posix.User.ByteString
+# else
+import System.Posix.User
+# endif
+  (GroupEntry, UserEntry)
 
 import TextShow.TH (deriveTextShow)
 

--- a/tests/Instances/System/Posix.hs
+++ b/tests/Instances/System/Posix.hs
@@ -29,7 +29,12 @@ import           Prelude.Compat
 
 import           System.Posix.DynamicLinker (RTLDFlags(..), DL(..))
 import           System.Posix.Process (ProcessStatus(..))
-import           System.Posix.User (GroupEntry(..), UserEntry(..))
+# if MIN_VERSION_unix(2,8,0)
+import           System.Posix.User.ByteString
+# else
+import           System.Posix.User
+# endif
+                   (GroupEntry(..), UserEntry(..))
 
 import           Test.QuickCheck (Arbitrary(..), arbitraryBoundedEnum)
 import           Test.QuickCheck.Instances ()

--- a/tests/Instances/Trace/Hpc.hs
+++ b/tests/Instances/Trace/Hpc.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP                #-}
 {-# LANGUAGE DataKinds          #-}
 {-# LANGUAGE StandaloneDeriving #-}
 {-# LANGUAGE TemplateHaskell    #-}
@@ -31,10 +32,12 @@ import           Trace.Hpc.Tix (Tix(..), TixModule(..))
 import           Trace.Hpc.Util (HpcPos, Hash, toHpcPos)
 
 $(Generics.deriveAll0 ''BoxLabel)
+$(Generics.deriveAll0 ''HpcPos)
+#if !(MIN_VERSION_hpc(0,6,2))
 $(Generics.deriveAll0 ''Tix)
 $(Generics.deriveAll0 ''TixModule)
-$(Generics.deriveAll0 ''HpcPos)
 $(Generics.deriveAll0 ''Hash)
+#endif
 
 instance Arbitrary Mix where
     arbitrary = Mix <$> arbitrary <*> arbitrary <*> arbitrary

--- a/tests/Spec/System/PosixSpec.hs
+++ b/tests/Spec/System/PosixSpec.hs
@@ -25,7 +25,12 @@ import Spec.Utils (matchesTextShowSpec)
 
 import System.Posix.DynamicLinker (RTLDFlags, DL)
 import System.Posix.Process (ProcessStatus)
-import System.Posix.User (GroupEntry, UserEntry)
+# if MIN_VERSION_unix(2,8,0)
+import System.Posix.User.ByteString
+# else
+import System.Posix.User
+# endif
+  (GroupEntry, UserEntry)
 
 import Test.Hspec (describe)
 

--- a/text-show-instances.cabal
+++ b/text-show-instances.cabal
@@ -161,7 +161,7 @@ library
     build-depends:     Win32                 >= 2.3.1.1 && < 2.13
   else
     build-depends:     terminfo              >= 0.4     && < 0.5
-                     , unix                  >= 2.7     && < 2.8
+                     , unix                  >= 2.7     && < 2.9
 
 test-suite spec
   type:                exitcode-stdio-1.0
@@ -279,4 +279,4 @@ test-suite spec
     build-depends:     Win32                 >= 2.3.1.1 && < 2.13
   else
     build-depends:     terminfo              >= 0.4     && < 0.5
-                     , unix                  >= 2.7     && < 2.8
+                     , unix                  >= 2.7     && < 2.9

--- a/text-show-instances.cabal
+++ b/text-show-instances.cabal
@@ -125,13 +125,13 @@ library
                        TextShow.System.Posix
   other-modules:       TextShow.Utils
   build-depends:       aeson                 >= 2.0.3   && < 2.2
-                     , base                  >= 4.9     && < 4.18
+                     , base                  >= 4.9     && < 4.19
                      , base-compat           >= 0.10    && < 1
                      , bifunctors            >= 5.2     && < 6
                      , binary                >= 0.8.3   && < 0.9
                      , containers            >= 0.5.7.1 && < 0.7
                      , directory             >= 1.3     && < 1.4
-                     , ghc-boot-th           >= 8.0     && < 9.5
+                     , ghc-boot-th           >= 8.0     && < 9.7
                      , haskeline             >= 0.7.3   && < 0.9
                      , hpc                   >= 0.6     && < 0.7
                      , old-locale            >= 1       && < 1.1
@@ -141,7 +141,7 @@ library
                      , scientific            >= 0.3.7   && < 0.4
                      , semigroups            >= 0.16.2  && < 1
                      , tagged                >= 0.4.4   && < 1
-                     , template-haskell      >= 2.11    && < 2.20
+                     , template-haskell      >= 2.11    && < 2.21
                      , text                  >= 0.11.1  && < 2.1
                      , text-short            >= 0.1     && < 0.2
                      , text-show             >= 3.4     && < 4
@@ -233,14 +233,14 @@ test-suite spec
                        Spec.System.Console.TerminfoSpec
                        Spec.System.PosixSpec
   build-depends:       aeson                 >= 2.0.3   && < 2.2
-                     , base                  >= 4.9     && < 4.18
+                     , base                  >= 4.9     && < 4.19
                      , base-compat           >= 0.10    && < 1
                      , bifunctors            >= 5.5.5   && < 6
                      , binary                >= 0.8.3   && < 0.9
                      , containers            >= 0.5.7.1 && < 0.7
                      , directory             >= 1.3     && < 1.4
                      , generic-deriving      >= 1.9     && < 2
-                     , ghc-boot-th           >= 8.0     && < 9.5
+                     , ghc-boot-th           >= 8.0     && < 9.7
                      , ghc-prim
                      , haskeline             >= 0.7.3   && < 0.9
                      , hpc                   >= 0.6     && < 0.7
@@ -253,7 +253,7 @@ test-suite spec
                      , random                >= 1.0.1   && < 1.3
                      , tagged                >= 0.4.4   && < 1
                      , scientific            >= 0.3.7   && < 0.4
-                     , template-haskell      >= 2.11    && < 2.20
+                     , template-haskell      >= 2.11    && < 2.21
                      , text-short            >= 0.1     && < 0.2
                      , text-show             >= 3.10    && < 4
                      , text-show-instances

--- a/text-show-instances.cabal
+++ b/text-show-instances.cabal
@@ -78,7 +78,9 @@ tested-with:         GHC == 8.0.2
                    , GHC == 8.8.4
                    , GHC == 8.10.7
                    , GHC == 9.0.2
-                   , GHC == 9.2.2
+                   , GHC == 9.2.7
+                   , GHC == 9.4.4
+                   , GHC == 9.6.1
 extra-source-files:  CHANGELOG.md, README.md
 cabal-version:       >=1.10
 

